### PR TITLE
Add an option to manually set `tpipeline_size`

### PR DESCRIPTION
**Pull request by WilliamHsieh**
The problem with `tpipeline#util#set_size()` is that when the `maxwidth` of `nvim_eval_statusline` is too small, `tpipeline` is not able to detect the `split_pattern` properly, resulting in a weird render of the embedded statusline.
And in most cases, truncation of the statusline is done through configuring statusline plugins, so I add the option for the users to manually set the rendered statusline size `tpipeline_size`.
The behavior remained the same when the option is left empty.

### CODE REVIEW

**Comment by vimpostor** [comment]
> The problem is that when the `maxwidth` of `nvim_eval_statusline` is too small

I don't understand the problem scenario, as `maxwidth` is always the width of the terminal, so it should just work out:

1. If `maxwidth` is larger than the statusline, the statusline is rendered in its entirety without artifacts
2. If `maxwidth` is smaller than the fully expanded statusline, then neovim will itself truncate the statusline. It is true that splitting based on split_pattern will not work, but since the statusline is exactly as long as the terminal, the user won't be able to tell that the entire statusline is rendered from the left side in its entirety (and right side is empty)

I explained this idea more clearly in 617c649e2e71c7a33e428d61284e0653083fe890.

Maybe you just need to set your tmux `set -g status-left-length` to a much larger value?

If not, then I am happy to accept this patch, but it would be nice if you could upload a screenshot of a broken statusline example together with the output of `:echo tpipeline#debug#info()` while it is broken.

**Comment by WilliamHsieh** [comment]
> If maxwidth is larger than the statusline, the statusline is rendered in its entirety without artifacts
If maxwidth is smaller than the fully expanded statusline, then neovim will itself truncate the statusline. It is true that splitting based on split_pattern will not work, but since the statusline is exactly as long as the terminal, the user won't be able to tell that the entire statusline is rendered from the left side in its entirety (and right side is empty)

I can confirm that all of these works properly when `status-left-length` is large enough.
However, when the statusline contains tmux-specific escape sequence (e.g. tmux format: #{?pane_in_mode,foo,bar} ), the statusline is usually long enough to be truncated.
In this case, if the tmux escape sequence is truncated in the middle will result in a weird render of the embedded statusline.

For example, I'm setting the leftmost component as
```lua
-- line 54
cur_mode = "#{?client_prefix,Prefix,#{?pane_in_mode,Copy,#{?pane_synchronized,Sync," .. cur_mode .. "}}}"
```
however, due to the truncation, only the `pane_in_mode` part (highlighted in pink) is properly accepted by tmux (resulting in `<x,Copy}` in the lower left corner), thus the rest of the statusline is not properly rendered (shorter than expected).

Screenshot 2023-01-26 at 14 41 01

After this PR (with `tpipeline_size` set to 200)
Screenshot 2023-01-26 at 14 49 21

I also agree in https://github.com/vimpostor/vim-tpipeline/commit/617c649e2e71c7a33e428d61284e0653083fe890 you say that 
> The clean solution would be to have neovim somehow return the location of the split

but I don't know how to achieve this 😅, so this is what I come up with in this PR, providing an option for users who have the need to manually set the render length of the statusline.

**Comment by vimpostor** [comment]
OK I understand the problem now, but is there any reason why you are not using the manual embed step, i.e. you could add the tmux specific part of the statusline in the tmux config like this:

```tmux
set -g status-left '#{?client_prefix,Prefix,#{?pane_in_mode,Copy,#{?pane_synchronized,Sync,}}}#(cat #{socket_path}-\#{session_id}-vimbridge)'
```

This would solve the problem as well and it would be even better, because even with your patch the size calculations by neovim are off because it thinks the length of `#{?client_prefix,Prefix,#{?pane_in_mode,Copy,#{?pane_synchronized,Sync,}}}` is **74**, when in reality it usually is **0**, after it gets expanded by tmux.

So neovim might start truncating other components because it thinks your string needs that much space, when in reality it doesn't need any space. And this is the real source of the problem: Tmux specific sections of the statusline should _usually_ not be configured in neovim, as it will confuse neovim's length estimations.


That being said, I am still happy to accept your patch, as I think there might be usecases where one still wants to configure tmux specific statusline components from neovim's side (e.g. dynamically using highlight groups from neovim for some tmux component), however I would prefer this option to not be documented, as it usually is only needed in very specific edge cases, so documenting it will make it more likely to be misused, and we might find a better workaround in the future, so I don't want users to rely on this option too much.

Also setting `g:tpipeline_size` to a very large value might create a performance problem, as neovim needs to layout all that space with statusline content, although something in the ballpark of `200` is still fine probably.

> > The clean solution would be to have neovim somehow return the location of the split

> but I don't know how to achieve this :sweat_smile: 

Returning the location of the split would still not be enough to solve the problem unfortunately, in fact I have a PoC patch lying around on my disk that implements returning the split location in neovim, but we would still need to know the length of the terminal window to avoid neovim truncating the statusline if it is opened in a tmux vertical split pane that is much smaller than the terminal width.


tl;dr: If you remove the documentation for this workaround option, I will merge your patch.

**Comment by WilliamHsieh** [comment]
> OK I understand the problem now, but is there any reason why you are not using the manual embed step, i.e. you could add the tmux specific part of the statusline in the tmux config like this (alternatively you could also use g:tpipeline_embedopts):

Because I set up the mode component (e.g. Copy) to also be reactive to the vim mode, I'm not sure if there's an option to split the embedded component into smaller pieces (so I can fine-tune the tmux-specific part in my tmux config).

> (e.g. dynamically using highlight groups from neovim for some tmux component)

This is part of the reason I'm going this route. The main reason being I would like to have a different statusline in neovim and without neovim. And the mode component should react to the tmux prefix and vim mode if possible.
I'll remove the document. Thank's again for your support :)

**Comment by vimpostor** [comment]
> The main reason being I would like to have a different statusline in neovim and without neovim.

This could be solved by combining `g:tpipeline_embedopts` and `g:tpipeline_restore`.

Thanks again for the patch. And btw I really like your statusline rice, combining the tmux and vim mode in one component is a neat idea!

Let's see if we can fix this problem a bit more robust in the future (I already have an idea for allowing to add components after `nvim_eval_statusline` is evaluated, but need to come up with a design that is easily customizable and allows to add components inbetween).
